### PR TITLE
Allow statically configurable admin IDs

### LINKED ISSUE #2627: Consider introducing a static "admin identity" configurable in SPIRE Server
Currently, if a workload wants to access privileged SPIRE Server APIs, it must be registered with SPIRE Server as an admin workload. This is useful for some scenarios but cumbersome for others. For example, when operating as part of a larger PKI, it may be desirable to authorize callers for these APIs even if they have identities that are not directly managed by the local cluster. This can also be the case in environments that have orchestrated bootstrap logic, in which a handful of identities are minted and injected as part of the bootstrap process.

This could be as simple as a new SPIRE Server configurable e.g. `admin_ids = []` where one or more SPIFFE IDs can be statically supplied. For now, we should validate that these IDs reside in the same trust domain as our own, however in the future we may wish to allow federated authentication here as well.

### PR DESCRIPTION
**Pull request by azdagron**
Adds a new server configurable, `admin_ids`, where a set of fixed admin IDs can be configured. Callers presenting X509-SVIDs with one of those IDs will be granted admin privileges.

Fixes #2627

### CODE REVIEW

**Comment by radoslav-tomov** [comment]
This looks interesting. I've found the topic of bootstrapping admin identities to be somewhat lacking right now. My approach is a bootstrap script in which I've wrapped the server startup procedure, which handle the creation of admin identity. The script uses the local socket.api, which has no auth&autz checks. Long story short , the approach gives my finer control over the admin identity and I can specify and fine tune the selectors for the attestation process.
Something like:

```bash
    ./spire-server entry create \
        -admin \
        -parentID ${AGENT_SPIFFE_ID} \
        -spiffeID ${ADMIN_SPIFFE_ID} \
        -selector k8s:ns:my-namespace \
        -selector k8s:sa:spire-admin-sa
```

Maybe I don't fully comprehend the change, but I'm missing the minting of the admin SVID. Right now the change handles the AuthZ part.  The coding as well seems to be introducing a special handling for the configured admin entries, but there has already been a logic, which checks if an entry is actually defined as admin. 

Again I'm not that into in the coding of the project, but from usability/maintenance point of view, I would find it way more useful if we generalise it a bit, and reuse the existing coding. Maybe add a list of entries in the configuration, which on startup are added via the existing APIs. This way we will be way more versatile. 

```yaml
    identities:
        - parentID: "spiffe://example.org/my/agent" # optional
        - spiffeID: "spiffe://example.org/my/admin"
        selectors:
        - "k8s:ns:my:namespace"
        isAdmin:true
        isNode: false       
```

Let me know what you think . :)

**Comment by amartinezfayo** [comment]
@radoslav-tomov Thanks for commenting on this PR! I personally have some opinions and I would like to contribute to this conversation.
> Maybe I don't fully comprehend the change, but I'm missing the minting of the admin SVID.

I think that the motivating issue is more about being able to statically tell to a specific SPIRE Server which SPIFFE IDs should be considered admin IDs and not about a new way to add entries and mint SVIDs.

> Maybe add a list of entries in the configuration, which on startup are added via the existing APIs.

Introducing statically configured registrations poses some challenges, opening the possibility of having multiple views of the set of registrations across the running SPIRE servers, which presents a challenge to maintain the consistency of the data. Keeping consistency of registration entries is security sensitive, and particularly registrations with the admin bit set.

Due to that, I'm personally hesitant to try to go down that path and I feel more inclined to take the approach exposed in this PR.

**Comment by evan2645** [comment]
> Maybe I don't fully comprehend the change, but I'm missing the minting of the admin SVID. Right now the change handles the AuthZ part. 
> ...
>> Introducing statically configured registrations poses some challenges, opening the possibility of having multiple views of the set of registrations across the running SPIRE servers, which presents a challenge to maintain the consistency of the data.

I think @amartinezfayo is right and that this is the biggest challenge with what @radoslav-tomov proposes. That said, I do think we will need some functionality like that in the future ... I previously opened https://github.com/spiffe/spire/issues/2626 to discuss that.

This feature is intended for use in circumstances where the admin identity itself is minted elsewhere (in the PKI but outside SPIRE control, or on upstream/downstream nested SPIRE Servers, etc).

> Let me know what you think . :)

If you're using Kubernetes, you can have a look over here and see if it makes things easier for you https://github.com/spiffe/spire-controller-manager/blob/main/api/v1alpha1/clusterspiffeid_types.go#L69

**Comment by radoslav-tomov** [comment]
@amartinezfayo, @evan2645  thanks for the context, i think I understand now the use case.

> Introducing statically configured registrations poses some challenges, opening the possibility of having multiple views of the set of registrations across the running SPIRE servers, which presents a challenge to maintain the consistency of the data. Keeping consistency of registration entries is security sensitive, and particularly registrations with the admin bit set.

True story. Consistency can be an issue indeed. The current approach has even more challenging complications:
* how are we going to revoke admin access ?
* how are we going to list the entities, which have admin access ? 

Current approach is one promoting delegation. Relying on externally issued x509 certificates is a bit dangerous as certificates need to be issued, distributed, rotated and etc. -  things which we know are hard, often tedious and prone to human errors.  This direction increases the operations complexity of the product. I would like to see the opposite prioritised.    

For me granting admin access is the best place to eat our own dog food. Delegating it looks to me as step in the wrong direction.

**Comment by evan2645** [comment]
> For me granting admin access is the best place to eat our own dog food. Delegating it looks to me as step in the wrong direction.

You will still be able to directly issue and manage admin identities the same way you do today, all of that will continue to be supported. The patch in this PR simply makes it easier to delegate the creation/management of an admin identity, but not at the cost of all inclusive options.

Also as I mentioned earlier, the kind of thing you were originally asking for is something that is also under consideration (a way to inject bootstrap entries)

**Comment by radoslav-tomov** [comment]
Thanks. I seem to be mixing the 2 concerns :)

**Comment by MarcosDY** [comment]
how can users know the statically configured admins? maybe we can add a log about configured admin IDs when spire-server starts?

**Comment by MarcosDY** [comment]
LGTM!

File: conf/server/server_full.conf [hunk 1]
```diff
@@ -3,6 +3,11 @@
 
 # server: Contains core configuration parameters.
 server {
+    # admin_ids: SPIFFE IDs that, when present in a caller's X509-SVID, grant
+    # that caller admin privileges. The admin IDs must reside in the same
+    # trust domain as the server.
```
Comment (amoore877) [inline conf/server/server_full.conf:8]: > The admin IDs must reside in the same trust domain as the server.
Maybe for clarity on intention something like: "The admin IDs must reside in the same trust domain as the server, but need not have a corresponding registration entry with the server."

Comment (azdagron) [reply conf/server/server_full.conf:8]: Cool cool cool, will fix.

**Review by amartinezfayo — APPROVED** [review]
Thank you @azdagron!

**Pull request merged**